### ISSUE TITLE
cf push fails in staging process

### ISSUE THREAD

**Issue by msathe-tech** [cloudfoundry/cf-for-k8s#210]
### Summary

_In your own words, describe the issue._
Setup CF on TKG/AWS. CF push fails in staging process.

### Reproduction Steps
[cf-for-k8s sha hyperlink here]
cf push test-node-app -p tests/smoke/assets/test-node-app

Error
   Running 'npm install'
   node_modules      -> Not found
   npm-cache         -> Not found
   package-lock.json -> Not found
   Selected NPM build process: 'npm install'
   Executing build process
   Completed in 783ms
   Configuring environment
   **NPM_CONFIG_LOGLEVEL   -> error**
   NPM_CONFIG_PRODUCTION -> true
   PATH                  -> $PATH:/layers/org.cloudfoundry.npm/modules/node_modules/.bin
   Reusing layers from image 'index.docker.io/msathepivotal/9c2c4b22-2d2b-41f8-831e-a955f48889bd@sha256:a1e65a9cbca362f7b1fec2c7342bbe5d006ff5a7d4ecf16558e47ab7e87dcf0f'
   Reusing layer 'launcher'
   Reusing layer 'org.cloudfoundry.node-engine:node'
   Reusing layer 'org.cloudfoundry.npm:modules'
   Reusing 1/1 app layer(s)
   Reusing layer 'config'
   *** Images (sha256:8b59fc5890e5c28190eeb3ea89b95da7ef6a360cacbf0826d1dfca3035048247):
   msathepivotal/9c2c4b22-2d2b-41f8-831e-a955f48889bd
   index.docker.io/msathepivotal/9c2c4b22-2d2b-41f8-831e-a955f48889bd:b2.20200526.184838
   Adding cache layer 'org.cloudfoundry.node-engine:node'
   Adding cache layer 'org.cloudfoundry.npm:npm-cache'
   Build successful

Waiting for app to start...

**Stats unavailable: Stats server temporarily unavailable.
FAILED**

[Please include the `kapp deploy...` command, including all `config-optional` files]

[IaaS name(s) here (GKE, AKS, EKS, minikube, kind, etc)]

output of `ytt --version ; kapp --version ; kubectl version`:
[paste output here]

[Special configuration here]

_What steps/actions led to the issue?_

_It's helpful to include snippets of the error response or logs output_

### Expected behavior

_What was the expected result?_

**Comment by acosta11** [comment]
Hello,

This looks like the symptoms of https://github.com/cloudfoundry/cf-for-k8s/issues/70 . Would you happen to be using cf cli v6 and have not enabled the `diego-docker` feature flag on your cf foundation? Does the app push still fail even if you enable that feature flag?

If that does not work, do you happen to have the version/commit sha of cf-for-k8s that you are running to aid in further troubleshooting?

Thanks,
Andrew and @jpalermo

**Comment by msathe-tech** [comment]
cf version is 6.5
For diego-docker there is no instruction in the README. Could you please update the docs here - https://github.com/cloudfoundry/cf-for-k8s/blob/master/docs/deploy.md

**Comment by msathe-tech** [comment]
Im following the latest code from https://github.com/cloudfoundry/cf-for-k8s

**Comment by msathe-tech** [comment]
I tried now with - 
`cf enable-feature-flag diego_docker
`
Still the same error.

**Comment by acosta11** [comment]
> cf version is 6.5
For diego-docker there is no instruction in the README. Could you please update the docs

> Im following the latest code from https://github.com/cloudfoundry/cf-for-k8s

Interesting.  The request to try the feature flag was more of a potential symptom of using an older version of cf-for-k8s. When we resolved #70 we removed all references to the feature flag from docs. Before adding any references back to the docs, we might want to confirm whether it's actually related to this issue.

Maybe this is an issue on tkg/aws we haven't yet encountered because our ci currently only tests on gke and aks.

A few things to check next:
- Are there any additional logs in the cf logs --recent output?
- Does Eirini schedule an app pod in your cf-workloads namespace?
- Are there any weird logs in capi-kpack-watcher after the succesful staging/build?
- Are there any logs in the capi api-server pods indicating an error sending a request to Eirini for scheduling the app pods?

Thanks,
Andrew and @jpalermo

**Comment by msathe-tech** [comment]
Interestingly, the app is showing up in `cf a `and can be accessed. Will really help to include a few debugging instructions in the docs

**Comment by acosta11** [comment]
> cf version is 6.5

Additionally, does the error persist even when using cf cli v7?

**Comment by acosta11** [comment]
Synced up with the team and found that this may also be an issue with the environment missing a metrics-server.

```
Waiting for app to start...

Stats unavailable: Stats server temporarily unavailable.
FAILED
```

The cf push command does not evaluate the app as started until it successfully scrapes metrics from the pod, which requires metric-server. Does your cluster have this installed?

From the `deploy.md` docs:
> Supports metrics-server
Most IaaSes come with metrics-server, but if yours does not come with it or if you're using kind, then you may want to run something like
  `kubectl apply -f https://github.com/kubernetes-sigs/metrics-server/releases/download/v0.3.6/components.yaml`

**Comment by msathe-tech** [comment]
Nope, it wasn't there. Looks like TKG on AWS cluster created by TMC is not setup with TAS abstraction prereqs. It will really help if README is updated to highlight what is required for TKG on AWS.

**Comment by acosta11** [comment]
> Nope, it wasn't there. Looks like TKG on AWS cluster created by TMC is not setup with TAS abstraction prereqs.
Thanks for the confirmation.

> It will really help if README is updated to highlight what is required for TKG on AWS.

I raised this with our PM yesterday to see if we could prioritize ironing out the details of AWS installation and operation. Currently, we've only allocated time investigating and validating GKE and AKS in CI to focus on integrating additional features.

I'm going to leave this open for now and circle back tomorrow in case we missed anything.

Thanks,
Andrew

**Comment by acosta11** [comment]
Hello,

To close the loop on this issue, we merged #213 last week and as it appears stable, we expect issues related to missing metrics server are now resolved. I'm closing this issue, but feel free to reopen it if this persists on cf-for-k8s at https://github.com/cloudfoundry/cf-for-k8s/commit/3b7e654afeb7e9c3a99d865f63437e9fbf9e9f9f or later.

Thanks,
Andrew

**Comment by jamespollard8** [comment]
Sweet - thanks Costa! My understanding is that we now have a good solution and better documentation for setting up clusters that don't come with a pre-installed metrics-server.

@msathe-tech we love feedback - please keep it coming!
